File: front-end/src/lib/api/prospectProfiles.ts
import { z } from 'zod';
import { apiClient, ApiClientError } from './client';

/**
 * Zod schema mirroring the backend `ProspectProfileResponse` shape.
 * Backed by the `prospect_profiles` table; currently the only source is
 * Dane Brugler's "the-beast-2026" guide.
 */
export const ProspectProfileSchema = z.object({
	id: z.string().uuid(),
	player_id: z.string().uuid(),
	source: z.string(),
	grade_tier: z.string().nullable(),
	overall_rank: z.number().int().nullable(),
	position_rank: z.number().int(),
	year_class: z.string().nullable(),
	birthday: z.string().nullable(), // YYYY-MM-DD
	jersey_number: z.string().nullable(),
	height_raw: z.string().nullable(),
	nfl_comparison: z.string().nullable(),
	background: z.string().nullable(),
	summary: z.string().nullable(),
	strengths: z.array(z.string()),
	weaknesses: z.array(z.string()),
	college_stats: z.unknown().nullable(),
	scraped_at: z.string(), // YYYY-MM-DD
});

export type ProspectProfile = z.infer<typeof ProspectProfileSchema>;

/**
 * Lightweight summary returned by the bulk list endpoint. Drops the heavy
 * prose fields so the player list / prospects page can render grade tier
 * badges without paying the cost of background, summary, strengths, etc.
 */
export const ProspectProfileSummarySchema = z.object({
	player_id: z.string().uuid(),
	source: z.string(),
	grade_tier: z.string().nullable(),
	overall_rank: z.number().int().nullable(),
	position_rank: z.number().int(),
	nfl_comparison: z.string().nullable(),
});

export type ProspectProfileSummary = z.infer<typeof ProspectProfileSummarySchema>;

export const prospectProfilesApi = {
	/**
	 * Fetch the latest prospect profile for a player. Returns `null` when
	 * no profile exists (404), so callers can render conditionally.
	 */
	async getByPlayer(playerId: string): Promise<ProspectProfile | null> {
		try {
			return await apiClient.get(`/players/${playerId}/profile`, ProspectProfileSchema);
		} catch (e) {
			if (e instanceof ApiClientError && e.status === 404) {
				return null;
			}
			throw e;
		}
	},

	/**
	 * Fetch all profile summaries from a given source (default: the-beast-2026).
	 * Returns a Map keyed by player_id for O(1) lookup at render time.
	 */
	async loadSummariesBySource(
		source = 'the-beast-2026'
	): Promise<Map<string, ProspectProfileSummary>> {
		const list = await apiClient.get(
			`/prospect-profiles?source=${encodeURIComponent(source)}`,
			z.array(ProspectProfileSummarySchema)
		);
		const map = new Map<string, ProspectProfileSummary>();
		for (const entry of list) {
			map.set(entry.player_id, entry);
		}
		return map;
	},
};

File: back-end/crates/api/src/handlers/prospect_profiles.rs
use axum::extract::{Path, State};
use axum::Json;
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use utoipa::ToSchema;
use uuid::Uuid;

use crate::error::{ApiError, ApiResult};
use crate::state::AppState;

/// Public-facing shape of a `ProspectProfile`. Mirrors the domain model with
/// `serde_json::Value` for the JSONB columns so the frontend can render flexibly.
#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct ProspectProfileResponse {
    pub id: Uuid,
    pub player_id: Uuid,
    pub source: String,
    pub grade_tier: Option<String>,
    pub overall_rank: Option<i32>,
    pub position_rank: i32,
    pub year_class: Option<String>,
    pub birthday: Option<NaiveDate>,
    pub jersey_number: Option<String>,
    pub height_raw: Option<String>,
    pub nfl_comparison: Option<String>,
    pub background: Option<String>,
    pub summary: Option<String>,
    pub strengths: Vec<String>,
    pub weaknesses: Vec<String>,
    pub college_stats: Option<JsonValue>,
    pub scraped_at: NaiveDate,
}

impl From<domain::models::ProspectProfile> for ProspectProfileResponse {
    fn from(p: domain::models::ProspectProfile) -> Self {
        Self {
            id: p.id,
            player_id: p.player_id,
            source: p.source,
            grade_tier: p.grade_tier,
            overall_rank: p.overall_rank,
            position_rank: p.position_rank,
            year_class: p.year_class,
            birthday: p.birthday,
            jersey_number: p.jersey_number,
            height_raw: p.height_raw,
            nfl_comparison: p.nfl_comparison,
            background: p.background,
            summary: p.summary,
            strengths: p.strengths,
            weaknesses: p.weaknesses,
            college_stats: p.college_stats,
            scraped_at: p.scraped_at,
        }
    }
}

/// GET /api/v1/players/{player_id}/profile
///
/// Returns the most recently scraped prospect profile for the given player,
/// or 404 if no profile exists. Currently the only source is "the-beast-2026".
pub async fn get_player_profile(
    State(state): State<AppState>,
    Path(player_id): Path<Uuid>,
) -> ApiResult<Json<ProspectProfileResponse>> {
    let profile = state
        .prospect_profile_repo
        .find_latest_by_player(player_id)
        .await?
        .ok_or_else(|| {
            ApiError::NotFound(format!("No prospect profile found for player {}", player_id))
        })?;

    Ok(Json(profile.into()))
}
